fiftyone/core/collections.py

```diff
@@ -69,49 +71,97 @@ def registrar(func):
 aggregation = _make_registrar()
 
 
-class _AutosaveSample:
-    def __init__(self, batch, sample):
-        self._batch = batch
-        self._sample = sample
+class SaveContext(object):
+    """Context that saves samples from a collection according to a configurable
+    batching strategy.
+
+    Args:
+        sample_collection: a
+            :class:`fiftyone.core.collections.SampleCollection`
+        batch_size (None): the batching strategy to use. Can either be an
+            integer specifying the number of samples to save in a batch, or a
+            float number of seconds between batched saves
+    """
+
+    def __init__(self, sample_collection, batch_size=None):
+        if batch_size is None:
+            batch_size = 0.2
+
+        self.sample_collection = sample_collection
+        self.batch_size = batch_size
+
+        self._dataset = sample_collection._dataset
+        self._sample_coll = sample_collection._dataset._sample_collection
+        self._frame_coll = sample_collection._dataset._frame_collection
+
+        self._sample_ops = []
+        self._frame_ops = []
+        self._reload_parents = []
+
+        self._curr_batch_size = None
+        self._dynamic_batches = not isinstance(batch_size, numbers.Integral)
+        self._last_time = None
 
     def __enter__(self):
+        if self._dynamic_batches:
+            self._last_time = timeit.default_timer()
+
+        self._curr_batch_size = 0
         return self
 
-    def __exit__(self, exc_type, exc_value, exc_tb):
-        self._batch.save_sample(self._sample)
+    def __exit__(self, *args):
+        self._save_batch()
 
+    def save(self, sample):
+        """Registers the sample for saving in the next batch.
 
-class _AutosaveSampleBatch:
-    def __init__(self, sample_collection, frame_collection, batch_size=None):
-        self._sample_collection = sample_collection
-        self._frame_collection = frame_collection
-        self._batch_size = batch_size or 3
-        self._samples = []
+        Args:
+            sample: a :class:`fiftyone.core.sample.Sample` or
+                :class:`fiftyone.core.sample.SampleView`
+        """
+        if sample._in_db and sample._dataset is not self._dataset:
+            raise ValueError(
+                "Dataset context '%s' cannot save sample from dataset '%s'"
+                % (self._dataset.name, sample._dataset.name)
+            )
 
-    def save_sample(self, sample):
-        self._samples.append(sample)
-        if len(self._samples) >= self._batch_size:
-            self._save_samples()
+        sample_op, frame_ops = sample._save(deferred=True)
```

Comment:
I'm calling `_save()` here rather than waiting until `_save_batch()` is called to spread the computation across calls to `save()`. If the user is printing a progress bar for the *input* samples, this makes it smoother.